File: geeks-for-geeks/kotlin/greedy-algorithms/graphColoring.kt
import java.io.*
import java.util.*
import java.util.LinkedList

internal class Graph(private val V: Int) {
    private val adj: Array<LinkedList<Integer>>
    fun addEdge(v: Int, w: Int) {
        adj[v].add(w)
        adj[w].add(v)
    }

    fun greedyColoring() {
        val result = IntArray(V)
        Arrays.fill(result, -1)
        result[0] = 0
        val available = BooleanArray(V)
        Arrays.fill(available, true)
        for (u in 1 until V) {
            val it: Iterator<Integer> = adj[u].iterator()
            while (it.hasNext()) {
                val i: Int = it.next()
                if (result[i] != -1) available[result[i]] = false
            }
            var cr: Int
            cr = 0
            while (cr < V) {
                if (available[cr]) break
                cr++
            }
            result[u] = cr
            Arrays.fill(available, true)
        }
        for (u in 0 until V) System.out.println(
            "Vertex " + u + " ---> Color "
                    + result[u]
        )
    }

    companion object {
        fun main(args: Array<String?>?) {
            val g1 = Graph(5)
            g1.addEdge(0, 1)
            g1.addEdge(0, 2)
            g1.addEdge(1, 2)
            g1.addEdge(1, 3)
            g1.addEdge(2, 3)
            g1.addEdge(3, 4)
            System.out.println("Coloring of graph 1")
            g1.greedyColoring()
            System.out.println()
            val g2 = Graph(5)
            g2.addEdge(0, 1)
            g2.addEdge(0, 2)
            g2.addEdge(1, 2)
            g2.addEdge(1, 4)
            g2.addEdge(2, 4)
            g2.addEdge(4, 3)
            System.out.println("Coloring of graph 2 ")
            g2.greedyColoring()
        }
    }

    init {
        adj = arrayOfNulls<LinkedList>(V)
        for (i in 0 until V) adj[i] = LinkedList()
    }
}

File: geeks-for-geeks/js/greedy-algorithms/graphColoring.js
function addEdge(adj,v,w){
    adj[v] = [...adj[v],w]
    adj[w] = [...adj[w],v]
    // don't use adj[v].push(w)
    return adj
  }
  
  function graphColoring(adj,V){
     let result = Array(V).fill(-1)
     result[0] = 0
     let available = Array(V).fill(false)
     
     for(let u=1;u<V;u++){
       
       for(let i of adj[u]){
         if(result[i] != -1){
             available[result[i]] = true
         }
       }
  
       let cr=0
  
       while(cr < V){
          if(available[cr] === false)
          break;
          cr+=1
       }
  
       result[u] = cr
  
       for(let i of adj[u]){
         if(result[i] !== -1){
           available[result[i]] = false
         }
       }
  
     }
     
    for(let u=0;u<V;u++){
      console.log("Vertex :" + u + " ---> Color :" + result[u])
    }
  
  }
  
  var g1 = Array(5).fill([])
  g1 = addEdge(g1,0,1)
  g1 = addEdge(g1,0,2)
  g1 = addEdge(g1,1,2)
  g1 = addEdge(g1,1,3)
  g1 = addEdge(g1,2,3)
  g1 = addEdge(g1,3,4)
  graphColoring(g1,5)
  
  console.log("####################")
  
  let g2 = Array(5).fill([])
  g2 = addEdge(g2,0,1)
  g2 = addEdge(g2,0,2)
  g2 = addEdge(g2,1,2)
  g2 = addEdge(g2,1,4)
  g2 = addEdge(g2,2,4)
  g2 = addEdge(g2,4,3)
  
  graphColoring(g2,5)